packages/discovery-react-components/src/components/StructuredQuery/StructuredQuery.tsx

```diff
@@ -19,41 +19,50 @@ export interface StructuredQueryProps {
 
 const StructuredQuery: FC<StructuredQueryProps> = ({ messages = defaultMessages }) => {
   const mergedMessages = { ...defaultMessages, ...messages };
-  const [ruleRows, setRuleRows] = useState({ rows: [{ id: 0 }] });
-  const showAddRuleRowButton = ruleRows.rows.length < MAX_NUM_SIBLING_RULE_ROWS;
-
-  const handleAddRuleRowOnClick = () => {
-    const newRuleRowId = ruleRows.rows[ruleRows.rows.length - 1].id + 1;
-    const newRuleRow = { id: newRuleRowId };
-    setRuleRows({
-      ...ruleRows,
-      rows: ruleRows.rows.concat(newRuleRow)
-    });
-  };
+  const [structuredQuerySelection, setStructuredQuerySelection] = useState<
+    StructuredQuerySelection
+  >({
+    rows: [{ id: 0 }],
+    groups: []
+  });
+  const showAddRuleRowButton = structuredQuerySelection.rows.length < MAX_NUM_SIBLING_RULE_ROWS;
+  const showAddRuleGroupButton = structuredQuerySelection.groups.length < MAX_NUM_RULE_GROUPS;
 
   return (
     <div className={structuredQueryClass}>
-      <RuleGroupDropdown messages={mergedMessages} />
-      {ruleRows.rows.map(row => {
+      <RuleGroup
+        messages={mergedMessages}
+        groupId="top-level"
+        structuredQuerySelection={structuredQuerySelection}
+        setStructuredQuerySelection={setStructuredQuerySelection}
+      />
+      {structuredQuerySelection.groups.map(group => {
         return (
-          <RuleRow
+          <RuleGroup
             messages={mergedMessages}
-            rowId={row.id}
-            key={row.id}
-            setRuleRows={setRuleRows}
-            ruleRows={ruleRows}
+            groupId={group.id}
+            structuredQuerySelection={structuredQuerySelection}
+            setStructuredQuerySelection={setStructuredQuerySelection}
+            key={group.id}
           />
         );
       })}
       <div className={structuredQueryRulesButtonsClass}>
         {showAddRuleRowButton && (
-          <Button kind="ghost" renderIcon={Add16} onClick={handleAddRuleRowOnClick}>
-            {mergedMessages.addRuleRowText}
-          </Button>
+          <AddRuleRowButton
+            addRuleRowText={mergedMessages.addRuleRowText}
+            setStructuredQuerySelection={setStructuredQuerySelection}
+            structuredQuerySelection={structuredQuerySelection}
+            groupId="top-level"
```

Comment:
why not just `0`?

Reply:
I refactored with a data structure similar to what you recommended above (just with the `entities` part, except I'm calling it `groups`), so this is now `0` for top-level groups.